lib/connection.ex

```diff
@@ -291,6 +291,68 @@ defmodule Connection do
   """
   defcallback terminate(any, any) :: any
 
+  defmacro __using__(_) do
+    quote location: :keep do
+      @behaviour Connection
+
+      # The default implementations of init/1, handle_call/3, handle_info/2,
+      # handle_cast/2, terminate/2 and code_change/3 have been taken verbatim
+      # from Elixir's GenServer default implementation.
+
+      @doc false
+      def init(args) do
+        {:ok, args}
+      end
+
+      @doc false
+      def handle_call(msg, _from, state) do
+        # We do this to trick dialyzer to not complain about non-local returns.
+        case :random.uniform(1) do
+          1 -> exit({:bad_call, msg})
+          2 -> {:noreply, state}
+        end
+      end
+
+      @doc false
+      def handle_info(_msg, state) do
+        {:noreply, state}
+      end
+
+      @doc false
+      def handle_cast(msg, state) do
+        # We do this to trick dialyzer to not complain about non-local returns.
+        case :random.uniform(1) do
+          1 -> exit({:bad_cast, msg})
+          2 -> {:noreply, state}
+        end
+      end
+
+      @doc false
+      def terminate(_reason, _state) do
+        :ok
+      end
+
+      @doc false
+      def code_change(_old, state, _extra) do
+        {:ok, state}
+      end
+
+      @doc false
+      def connect(_, _state) do
+        exit(:bad_connect)
+      end
```

Comment:
The first argument is information for that connection attempt. Can you include that in the exit reason, `{:bad_connect, info}`. I am happy not to use the dialyzer trick here because the warning is valid - why use a connection if can't connect?